### DISCUSSION TITLE
Type-based deconstruction of instances without the need for a `Deconstruct` method

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#7160] [Language Ideas] [answered]
## Summary

This proposal enables the deconstruction of select properties that provide a visible get accessor, based on their type.

## Motivation

Some variables could need isolation out of the scope of the instance that provides those values from specific properties, which could increase maintainability in certain scenarios. Furthermore, it is often redundant to rely on the deconstructor to distinctly map specific properties of interest, discarding the rest with `_`, and using the order in which they appear in the parameters of the deconstructor.

## Design

All the types that provide visible properties can utilize this feature, including structs, classes, and records of their kind. This feature would specifically not rely on any present `Deconstruct` method, on the type itself, or an extension one.

The feature enables mapping distinctly available types to variables of the designated type. The target types of each variable must be explicitly provided on the variable declaration.

### Example

Consider the following record:
```csharp
public record Book(Author Author, string Title, Publisher Publisher, int Year);
```

If we were to only account for the author and the publisher of the book, using a standard deconstruction to store the values into variables would look like this:
```csharp
Book book = // get a book from somewhere
var (author, _, publisher, _) = book;
// use author and publisher
```

Despite the names of the locals that are defined here, the order of the deconstruction could significantly matter and is subject to confusing errors upon adjustment.

However, we notice that our two properties, the author and the publisher, are distinct in the type's visible parameters, and in this case in all scopes. This could enable the use of this feature, using the following syntax:

```csharp
var (Author author, Publisher publisher) = deconstruct book;
```

The `deconstruct` is a keyword in this context, and explicitly tells the compiler to lower the above piece of code into the following:
```csharp
var author = book.Author;
var publisher = book.Publisher;
```

This enables quick and local access to the instance's properties of interest, that could also be possibly overwritten accordingly, based on the business logic of the application.

**Comment by kyoyama-kazusa** [comment]
Hi.

I know this is very common, but I suggest you use another way to get them. Just define the type as a `record struct`:

```csharp
public record struct Book(Author Author, string Title, Publisher Publisher, int Year);
```

or a normal `struct`:

```csharp
public struct Book(Author author, string title, Publisher publisher, int year)
{
    public Author Author { get; } = author;
    public string Title { get; } = title;
    public Publisher Publisher { get; } = publisher;
    public int Year { get; } = year;
}
```

Then you can write:

```csharp
_ = book is { Author: var author, Publisher: var publisher };
```

Because `struct`s are not `null`, we can write this property pattern like an expression, and you won't get a compiler error to tell you this deconstruction may be failed if the variable `book` is null.

**Reply by Rekkonnect** [reply]
It is true that this is the closest you can get to this proposal, using the current language version, however it looks ugly and rather becomes an anti-pattern by abusing pattern matching. Plus, you still don't avoid the explicit matching of the property's name, without only regarding the type of the property.

**Reply by kyoyama-kazusa** [reply]
I also want to a better way to achieve this. However there is no simpler way to do with it. I don't think using `deconstruct` keyword is a good plan, because it can only be used for deconstruction, which is too narrow a case to be used; in addition, the word `deconstruct` is too long to write. For the aspect of intellisense suggestions, keywords start with letters d and e are `decimal`, `default` and `delegate`. I may be offen to write wrong keywords instead of `deconstruct`.

Just to get further consideration, suppose we can use a way to tell the compiler that we want to get the internal values without `null` or other kind of checking (like #4082), I think this is a good way to achieve this, so I give that proposal a thumb-up. If the keyword is only limited to deconstruction, I don't think this is good (only for the consideration of syntax chosen).

:)

**Reply by Rekkonnect** [reply]
I really agree, being probably restrictive about the flexibility of the syntax would prevent further expansion of deconstruction as a concept in general. Though, what I would really like is enabling this feature of matching against the explicit types of the properties, without knowing of the names of the properties, or without having to potentially duplicate the naming.

Also, very accurate comment on `deconstruct`. I very lazily chose a random keyword, without thinking about the exact syntax that would resolve potential ambiguities.

**Reply by HaloFour** [reply]
> Though, what I would really like is enabling this feature of matching against the explicit types of the properties, without knowing of the names of the properties, or without having to potentially duplicate the naming.

IMO that would make it much too restrictive.  It could never work with any types that happen to have multiple properties of the same type, which precludes the vast majority of types.  The name, if anything, is more important than the type.  Property-based deconstruction, like property-based matching, is nominal by design.

**Comment by HaloFour** [comment, marked as answer]
See: #4082

**Reply by Rekkonnect** [reply]
This is not exactly the same, this specific proposal binds to the type of the only property that matches exactly the provided type of the deconstructed local. However I see that proposal as a conflict of interest, where one way would be preferred over the other.

This means that one approach will be more likely to be accepted and considered for implementation.

**Reply by HaloFour** [reply]
Deconstruction is explicitly positional only, the types don't factor into it.  That is why you also can't overload the deconstructor based on type, only on arity.

That other proposal builds on pattern matching and deconstruction, which makes it much more applicable.  It is also already championed.
